### PR TITLE
Cms 1903

### LINKED ISSUE #1903: Create automated tests and add better error handling for USAJobs API calls
Related to https://github.com/fecgov/fec-cms/pull/1885

**Completion criteria:** 
- [ ] Add additional error handling and template language so that in the event that the USAJobs API is down or any of the queried endpoints are not returning  data we are alerted to this and direct users to the USAJobs page (preferrably a USAJobs page populated with all of our open jobs). From @PaulClark2: "USAJOBS is unavailable. Please visit usajobs.gov for more information."
- [ ] Add error handling for the [hiring-path get request](https://github.com/fecgov/fec-cms/blob/6556f39dc946cc4b1bfa2eaf1081955cc941f23e/fec/home/templatetags/open_jobs.py#L31) which is currently not coded to check to check for no data or null being returned
- [ ] Write automated tests that we can use to verify that our error handling works and that out app always returns matches what is posted on USAJobs

**Reported Error:**
On 7/27/18, the https://www.fec.gov/about/careers/ page threw an error due to some problems retrieving data from the USA JOBS api. (Here's the error in the logs: https://logs.fr.cloud.gov/goto/b52ce59f67ed40363196d6ed47b70d68). One theory is that we couldn’t get the job codes back for a short time:
`raise JSONDecodeError("Expecting value", s, err.value) from None`

### PR DESCRIPTION
**Pull request by qqss88**
## Summary (required)
some modules updated to add API error handling and unit tests:
1. open_jobs.py
2. test_usajobs.py
3. fec/fec/constants.py
4. fec/home/templates/partials/jobs.html

- Resolves #1903
# 1903
_Include a summary of proposed changes._

## Impacted areas of the application
List general components of the application that this PR will affect:

-  fec.gov --> about --> careers --> open positions

## Screenshots

- _Include a screenshot of the new/updated features in context (“in the wild”). If it is an interface or front end change, include both a Before and After screenshot._

## Related PRs
List related PRs against other branches:

branch | PR
------ | ------
fix/other_pr | [link]()
feature/other_pr | [link]()

## How to test
Include any information that may be helpful to the reviewer(s).
  This might inclide:
  - links to sample pages to test
  - Any local environmental setup that is different than the usual (env vars, API version to point to)
____

### CODE REVIEW

**Comment by johnnyporkchops** [comment]
@qqss88 , Everything works great. Great tests as well. As you suggested, I will push up code to add the link to the error message

**Comment by johnnyporkchops** [comment]
@JonellaCulmer , Here is a scrnsht of the error message now. It's a little weird because its inside the gray slab where job announcements (or 'no current positions...' text)  would normally go. If you are still not OK with it, just let us know and we will reconfigure the template.
screen shot 2018-10-30 at 9 14 06 pm

File: fec/home/views.py [hunk 1]
```diff
@@ -263,6 +263,7 @@ def serve_wagtail_doc(request, document_id, document_filename):
     doc = get_object_or_404(Document, id=document_id)
     return HttpResponseRedirect(doc.file.url)
 
+# TODO: undefined name 'paginator' in the function, need to be fixed.
```
Comment (lbeaufort) [inline fec/home/views.py]: Let's not leave a TODO in the code - we should fix the issue or remove the TODO.

File: fec/home/tests/test_usajobs.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    # but new code found, use original job data
+    @mock.patch("requests.get", side_effect=mocked_requests_get4)
+    def test_get_job4(self, mock_get):
+        logger.info("usajob test4:")
+        job_data = get_jobs()
+        self.assertEqual(job_data["jobData"][0]["open_to"], "some new code")
+        logger.info("pass.")
+
+
+if __name__ == "__main__":
```
Comment (lbeaufort) [inline fec/home/tests/test_usajobs.py]: Now that we're using `pytest`, is this needed?

https://github.com/fecgov/fec-cms/blob/8c1ed4deed05d52b9cd62872d51bec2149399e29/fec/home/tests/test_usajobs.py#L163-L164

File: fec/home/tests/test_usajobs.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        elif args[0] == CODES_URL:
+            return MockResponse(json.loads(USAJOBS_CODE_LIST), 200)
+        return MockResponse(None, 404)
+
+    # first test case: everything normal
+    @mock.patch("requests.get", side_effect=mocked_requests_get1)
+    def test_get_job1(self, mock_get):
+        logger.info("\n\nusajob test1:")
+        job_data = get_jobs()
+        # print(job_data)
```
Comment (lbeaufort) [inline fec/home/tests/test_usajobs.py]: we can take this out

File: fec/home/tests/test_usajobs.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+from home.templatetags.open_jobs import (
+    get_jobs,
+    USAJOB_SEARCH_ERROR,
+    JOB_URL,
+    CODES_URL,
+)
+from fec.constants import USAJOBS_CODE_LIST
+
+logger = logging.getLogger(__name__)
```
Comment (lbeaufort) [inline fec/home/tests/test_usajobs.py]: Now that we're using `pytest`, I don't think these logging statements will display - should we remove them?

File: fec/home/templatetags/open_jobs.py [hunk 5]
```diff
@@ -1,77 +1,111 @@
...
-    codes_url = 'https://data.usajobs.gov/api/codelist/hiringpaths'
-    querystring = {}
-    querystring['Organization']=settings.USAJOBS_AGENCY_CODE
-    querystring['WhoMayApply']=settings.USAJOBS_WHOMAYAPPLY
+    """
+    this function will query USAJOBS api and return all open FEC jobs.
+    if api call failed, a status error message will be displayed in the
+    jobs.html session in the career page.
+    it also query code list to update hirepath info. a hard-coded code list
+    is used for backup if query failed.
```
Comment (lbeaufort) [inline fec/home/templatetags/open_jobs.py]: This is a great approach!

File: fec/home/views.py (hunk 1 above)
Comment (qqss88) [reply fec/home/views.py]: I found out this issue when I setup the new environment for fec-cms. I just added a quick comment to mark it. This is not related to this ticket. I'll open another ticket for this one.

File: fec/home/tests/test_usajobs.py (hunk 2 above)
Comment (qqss88) [reply fec/home/tests/test_usajobs.py]: i think people still have the option to run this test as a standard python unit test or django unit test.

**Comment by johnnyporkchops** [comment]
**One more screenshots for the reconfigured template to handle the display of error message:**
(Retested to make sure everything else renders as expected with changed template logic)

screen shot 2018-11-05 at 6 28 50 pm

File: fec/home/views.py (hunk 1 above)
Comment (qqss88) [reply fec/home/views.py]: created new issue #2493 for this. TODO removed.

**Comment by JonellaCulmer** [comment]
@johnnyporkchops Thanks for that screenshot. It doesn't look like that when I ran it locally. Not sure where the disconnect lives. Also, can we remove the colon after the "Oops!"?

File: fec/home/tests/test_usajobs.py (hunk 2 above)
Comment (qqss88) [reply fec/home/tests/test_usajobs.py]: I am ok with removing them. There are more tests in the code base still having it - should we do a clean up?
./node_modules/node-gyp/gyp/pylib/gyp/common_test.py:  unittest.main()
./node_modules/node-gyp/gyp/pylib/gyp/input_test.py:  unittest.main()
./node_modules/node-gyp/gyp/pylib/gyp/generator/msvs_test.py:  unittest.main()
./node_modules/node-gyp/gyp/pylib/gyp/generator/ninja_test.py:  unittest.main()
./node_modules/node-gyp/gyp/pylib/gyp/generator/xcode_test.py:  unittest.main()
./node_modules/node-gyp/gyp/pylib/gyp/MSVSSettings_test.py:  unittest.main()
./node_modules/node-gyp/gyp/pylib/gyp/easy_xml_test.py:  unittest.main()

**Comment by johnnyporkchops** [comment]
If you want to test it locally you have to set the env vars:
USAJOBS_WHOMAYAPPLY =
USAJOBS_AGENCY_CODE =
USAJOBS_API_KEY =

File: fec/home/tests/test_usajobs.py (hunk 4 above)
Comment (qqss88) [reply fec/home/tests/test_usajobs.py]: removed logger from this test.

File: fec/home/tests/test_usajobs.py (hunk 2 above)
Comment (lbeaufort) [reply fec/home/tests/test_usajobs.py]: That's a great catch - I think a cleanup is a good idea.

**Comment by qqss88** [comment]
Did some cleanup as suggested by Laura. I think the python code is ready to be merged.

**Pull request merged**